### PR TITLE
Clustering update

### PR DESCRIPTION
**Pull request by trevorcampbell**
Major overhaul of the clustering section.

### CODE REVIEW

**Comment by ttimbers** [comment]
Some comments:
- Could we be a bit more explicit that clustering is a tool to answer exploratory statistical questions. And as such, the patterns we see are not extended to future cases (as in prediction) or generalized to the larger population (as in inference), but instead restricted to conclusions about the data in the dataset the clustering was performed on? But, as you already mention, useful for generating hypotheses/idea/new questions that could inform asking and answering additional predictive or inferential questions. Currently the term "exploratory" is only used in the overview section for the chapter.

**Comment by ttimbers** [comment]
Also, we might want to add a section discussing the `nstart` argument of `kmeans` as we use it in the tutorial (and maybe the worksheet)?

**Comment by ttimbers** [comment]
Last year, because R's outputs gives you tot sum of squares and between sum of squares, the students asked what these things were... However, another reason that might have come up was that a MDS students (and a TA for this course) gave this as a guest lecture. And he talked about these in his lecture slides... So my advice is either don't talk about these things anywhere, or explain them all...

File: 09-clustering.Rmd [hunk 1]
```diff
@@ -76,41 +122,323 @@ ggplot(data, aes(y = loyalty, x = csat, color = cluster)) +
...
+Then K-means consists of two major steps that attempt to minimize the
+sum of WSSDs over all the clusters, i.e. the *total WSSD*:
+
+1. **Center update:** Compute the center of each cluster.
+2. **Label update:** Reassign each data point to the cluster with the nearest center.
+
+These two steps are repeated until the cluster assignments no longer change.
+For example, in the customer data example from earlier, our initialization might look like this:
+<center>
+```{r 10-toy-kmeans-init, echo = FALSE, warning = FALSE, fig.height = 4, fig.width = 4.35, fig.cap = "Random initialization of labels."}
```
[the remark this answers is not in the document]
Comment (trevorcampbell) [reply 09-clustering.Rmd]: I don't see the package loading message you're referring to. Screenshot??

**Comment by trevorcampbell** [comment]
> Also, we might want to add a section discussing the `nstart` argument of `kmeans` as we use it in the tutorial (and maybe the worksheet)?

Done!

File: 09-clustering.Rmd [hunk 2]
```diff
@@ -76,41 +122,323 @@ ggplot(data, aes(y = loyalty, x = csat, color = cluster)) +
...
+p1
+```
+If we set K less than 3, then the clustering merges separate groups of data; this causes a large causing a large
+total WSSD, since the cluster center (denoted by an "x") is not close to any of the data in the cluster. On 
+the other hand, if we set K greater than 3, the clustering subdivides subgroups of data; this does indeed still 
+decrease the total WSSD, but by only a *diminishing amount*. If we plot the total WSSD versus the number of 
+clusters, we see that the decrease in total WSSD levels off (or forms an "elbow shape") when we reach roughly 
+the right number of clusters.
+<center>
+```{r 10-toy-kmeans-elbow, echo = FALSE, warning = FALSE, fig.height = 6, fig.width = 10, fig.cap = "Total WSSD for # clusters ranging from 1 to 9."}
```
[the remark this answers is not in the document]
Comment (trevorcampbell) [reply 09-clustering.Rmd]: I'll leave this as a TODO in the issues.

File: 09-clustering.Rmd [hunk 3]
```diff
@@ -76,41 +122,323 @@ ggplot(data, aes(y = loyalty, x = csat, color = cluster)) +
...
+In order to cluster data using K-means, we also have to pick the number of clusters, K.
+But unlike in classification, we have no data labels and cannot perform
+cross-validation with some measure of model prediction error.
+Further, if K is chosen too small, then multiple clusters get grouped together;
+if K is too large, then clusters get subdivided. In both cases, we will potentially miss
+interesting structure in the data. For example, take a look below at the K-means
+clustering of our customer satisfaction and loyalty data for a number of clusters
+ranging from 1 to 9.
+
+```{r 10-toy-kmeans-vary-k, echo = FALSE, warning = FALSE, fig.height = 10, fig.width = 12, fig.cap = "Clustering of the customer data for # clusters ranging from 1 to 9."}
```
[the remark this answers is not in the document]
Comment (trevorcampbell) [reply 09-clustering.Rmd]: Done for both. NB Lecture slides will still have Xs for this round; TODO for next round.

File: 09-clustering.Rmd [hunk 4]
```diff
@@ -76,41 +122,323 @@ ggplot(data, aes(y = loyalty, x = csat, color = cluster)) +
...
+    theme(legend.position = "none")
+print(plt_lbl)
+```
+</center>
+
+And the first three iterations of K-means would look like (each row corresponds to an iteration,
+where the left column depicts the center update, 
+and the right column depicts the reassignment of data to clusters): 
+<center>
+```{r 10-toy-kmeans-iter, echo = FALSE, warning = FALSE, fig.height = 4, fig.width = 8}
```
[the remark this answers is not in the document]
Comment (trevorcampbell) [reply 09-clustering.Rmd]: Done

File: 09-clustering.Rmd [hunk 5]
```diff
@@ -76,41 +122,323 @@ ggplot(data, aes(y = loyalty, x = csat, color = cluster)) +
...
+  xlim(c(0, 4)) +
+  ylim(c(6, 9)) +
+  geom_point(aes(y = mean(loyalty), x = mean(csat), colour="red", size=2.5)) +
+  theme(legend.position = "none")
+base
+```
 
-## K-means clustering in R
+The second step is to add up the squared distance between each point in the cluster and the cluster center.
+We use the usual distance formula that we learned about in the classification chapter.
```
[the remark this answers is not in the document]
Comment (trevorcampbell) [reply 09-clustering.Rmd]: done

File: 09-clustering.Rmd [hunk 6]
```diff
@@ -76,41 +122,323 @@ ggplot(data, aes(y = loyalty, x = csat, color = cluster)) +
...
 
-*note - when the advertisement pops up to register for this course, you can just click to ignore it (i.e., no need to sign up to watch the entire video)*
+The K-means algorithm is a procedure that groups data into K clusters.
+It starts with an initial clustering of the data, and then iteratively
+improves it by making adjustments to the assignment of data
+to clusters until it cannot improve any further. But how do we measure
+the "quality" of a clustering, and what does it mean to improve it? 
+In K-means clustering, we measure the quality of a cluster by its
+*within-cluster sum-of-squared-distances* (WSSD). Computing this involves two steps:
+We first compute the mean of each variable we are using to cluster the data
```
[the remark this answers is not in the document]
Comment (trevorcampbell) [reply 09-clustering.Rmd]: done!

**Comment by trevorcampbell** [comment]
> Could we be a bit more explicit that clustering is a tool to answer exploratory statistical questions. And as such, the patterns we see are not extended to future cases (as in prediction) or generalized to the larger population (as in inference), but instead restricted to conclusions about the data in the dataset the clustering was performed on? But, as you already mention, useful for generating hypotheses/idea/new questions that could inform asking and answering additional predictive or inferential questions. Currently the term "exploratory" is only used in the overview section for the chapter.

Clustering can actually help with prediction and other downstream tasks. Imagine a dataset that looks like a "cross" -- if we split the data using clustering and then do regression on each cluster separately, we might be a lot better at predicting things than if we tried to do a single regression on the whole dataset.

But since this course only uses clustering for exploration, I added a remark about that at the beginning.

**Pull request merged**